### ISSUE TITLE
airspeed-0.5.2dev_20150417 seems to not always be able to find variables in namespace.

### ISSUE THREAD

**Issue by ghuls** [purcell/airspeed#19]
airspeed-0.5.2dev_20150417 seems to not always be able to find variables in namespace.
airspeed-0.4.2dev_20141218 works fine.

The namespace dictionary contains the 'assembly' key.
In some cases everything worked fine (airspeed-0.5.2dev_20150417) when generating the output
files, in others (namespace dictionary contains some larger objects) it gave the following crash.
```
template_base_folder = os.path.dirname(template_filename)

with open(template_filename, 'r') as input_fh:
    template = airspeed.Template(input_fh.read())

output_fh.write(str(template.merge(namespace, loader=airspeed.CachingFileLoader(template_base_folder))))
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 70, in merge
    self.merge_to(namespace, output, loader)
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 81, in merge_to
    self.root_element.evaluate(fileobj, namespace, loader)
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 329, in evaluate
    return self.evaluate_raw(*args)
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 1175, in evaluate_raw
    self.block.evaluate(stream, namespace, loader)
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 329, in evaluate
    return self.evaluate_raw(*args)
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 1203, in evaluate_raw
    child.evaluate(stream, namespace, loader)
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 329, in evaluate
    return self.evaluate_raw(*args)
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 917, in evaluate_raw
    self.block.evaluate(stream, namespace, loader)
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 329, in evaluate
    return self.evaluate_raw(*args)
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 1203, in evaluate_raw
    child.evaluate(stream, namespace, loader)
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 329, in evaluate
    return self.evaluate_raw(*args)
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 1159, in evaluate_raw
    self.block.evaluate(stream, namespace, loader)
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 329, in evaluate
    return self.evaluate_raw(*args)
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 1203, in evaluate_raw
    child.evaluate(stream, namespace, loader)
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 335, in evaluate
    TemplateExecutionError(self, exc_info), exc_info[2])
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 329, in evaluate
    return self.evaluate_raw(*args)
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 698, in evaluate_raw
    value = self.expression.calculate(namespace, loader)
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 633, in calculate
    value = self.part.calculate(namespace, loader, global_namespace)
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 579, in calculate
    result = current_object[self.name]
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 213, in __getitem__
    return self.parent[key]
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 213, in __getitem__
    return self.parent[key]
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 213, in __getitem__
    return self.parent[key]
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 213, in __getitem__
    return self.parent[key]
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 213, in __getitem__
    return self.parent[key]
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 213, in __getitem__
    return self.parent[key]
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 213, in __getitem__
    return self.parent[key]
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 213, in __getitem__
    return self.parent[key]
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 213, in __getitem__
    return self.parent[key]
    
...
    
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 213, in __getitem__
    return self.parent[key]
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 213, in __getitem__
    return self.parent[key]
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 213, in __getitem__
    return self.parent[key]
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 213, in __getitem__
    return self.parent[key]
  File "/virtenv/local/lib/python2.7/site-packages/airspeed-0.5.2dev_20150417-py2.7.egg/airspeed/__init__.py", line 211, in __getitem__
    return dict.__getitem__(self, key)
TemplateExecutionError: Error in template '<string>' at position 1487-1498 in expression: ${assembly}
RuntimeError: maximum recursion depth exceeded
```

When reverting back to airspeed-0.4.2dev_20141218 (which I used before) everything works fine.

I tried to isolate a test case with the contents of namespace dictionary that causes the crash, but in
the isolated case it worked.

**Comment by purcell** [comment]
Hi - thanks for the report. This was probably due to 223337720b9c3a13d049187262235a8d49ea8af5 somehow, but all the tests are passing: if you can put together a reproducible example I'd love to fix this. Perhaps you could anonymise the actual template you're using so I can try it out?

**Comment by ghuls** [comment]
I was able to make a reproducible test case:

**airspeed_reduced_test.py**:
```python
import airspeed


def airspeed_reduced_test(loop_iterations):
    namespace = {'end': loop_iterations, 'assembly': 'hg19'}
    template_filename = 'reduced_test.vm'
    template_base_folder = '.'

    with open(template_filename, 'r') as input_fh:
        template = airspeed.Template(input_fh.read())

    return str(template.merge(namespace, loader=airspeed.CachingFileLoader(template_base_folder)))

print 'This succeeds:'
print '______________\n'
print airspeed_reduced_test(324)

print '\nThis fails:'
print '___________\n'
print airspeed_reduced_test(325)
```

**reduced_test.vm**:
```
#foreach($i in [1..$end])
    $assembly##
#end
```

**Comment by purcell** [comment]
Excellent, thanks very much. I'll try to get it fixed in the next few days.

**Comment by purcell** [comment]
Fixed in 0.5.3. It was a silly mistake on my part, and if it worked in the older version, it was probably just because the stack was shallow: more loop repetitions would likely have broken it. Thanks for your help!

**Comment by ghuls** [comment]
Thanks for fixing it.
